src/middlewared/middlewared/plugins/cluster_linux/ctdb_private.py

```diff
@@ -0,0 +1,232 @@
+        symlink_it = delete_it = False
+        if etc_file.exists():
+            if not etc_file.is_symlink():
+                # delete it since we're symlinking it
+                delete_it = True
+            else:
+                # means it's a symlink but not to the ctdb
+                # shared volume private ip file
+                if not etc_file.resolve() == ctdb_file:
+                    delete_it = True
+        else:
+            symlink_it = True
+
+        if delete_it:
+            try:
+                etc_file.unlink()
+            except Exception as e:
+                raise CallError(f'Failed deleting {etc_file} with error {e}')
+
+        if symlink_it:
+            try:
+                etc_file.symlink_to(ctdb_file)
+            except Exception as e:
+                raise CallError(f'Failed symlinking {etc_file} to {ctdb_file} with error {e}')
+
+        # ctdb documentation is _VERY_ explicit about
+        # how this file should be modified
+        if not delete:
+            # in the case of adding a node, it _MUST_ be
+            # added to the end of the file always
+            with open(PRI_IP_FILE, 'a') as f:
+                f.write(pri_ip + '\n')
+        else:
+            with open(PRI_IP_FILE) as f:
+                lines = f.read().splitlines()
+
+            # in the case of removing a node, a "#"
+            # _MUST_ be put in front of the entry in
+            # the file and the other lines _MUST_ be
+            # kept in the same order when writing it
+            # back out or ctdb will barf and fail to
+            # reload...
+            index = lines.index(pri_ip)
+            lines[index] = '#' + pri_ip
+
+            with open(PRI_IP_FILE, 'w') as f:
+                f.writelines(map(lambda x: x + '\n', lines))
+
+    @accepts(Dict(
+        'node_create',
+        IPAddr('ip'),
+    ))
+    @job(lock=PRI_LOCK)
+    async def create(self, job, data):
+        """
+        Add a ctdb private address to the cluster
+
+        `ip` is an IP v4/v6 address
+        """
+
+        schema_name = 'node_create'
+        verrors = ValidationErrors()
+
+        await self.middleware.call('ctdb.private.ips.common_validation', data['ip'], schema_name, verrors)
+        await self.middleware.call('ctdb.private.ips.update_file', data['ip'], verrors)
+
+        return data['ip']
+
+    @accepts(Str(id))
+    @job(lock=PRI_LOCK)
+    async def delete(self, job, id):
+        """
+        Delete a Private IP address from the ctdb cluster.
+
+        `id` is an IP v4/v6 address
+        """
+
+        schema_name = 'node_delete'
+        verrors = ValidationErrors()
+
+        address = (await self.get_instance(id))['id']
+        await self.middleware.call('ctdb.private.ips.common_validation', address, schema_name, verrors, True)
+        await self.middleware.call('ctdb.private.ips.update_file', address, verrors, True)
+
+        return address
```

Comment:
For consistency with CRUD API, we will probably also want an `update()` method that accepts an `id` / `pnn`. This can be a wrapper around your existing update_file() method.

Reply:
Spoke in DM and you're going to add a new `enabled` keyword to the `listnodes` ctdb command so that implementing a `do_update()` method makes a little more sense.